projs/shadow/shadow-engine/core/src/core/module-manager-v2.cpp

```diff
@@ -0,0 +1,216 @@
+        const std::string &symbolName = holder.descriptor.class_name + "_entry";
+
+        if (!assembly.lib->has_symbol(symbolName)) {
+            spdlog::error("❌ Could not find the entry for module \"{0}\"", holder.descriptor.id);
+            return;
+        }
+        try {
+            auto moduleInit = assembly.lib->get_function<std::shared_ptr<ShadowEngine::Module>()>(symbolName);
+
+            holder.module = moduleInit();
+        }
+        catch (std::exception &e) {
+            spdlog::error("❌ Error while running the entry for module \"{0}\" Error: {1}", holder.descriptor.id,
+                          e.what());
+            return;
+        }
+
+        holder.enabled = true;
+    }
+
+    void ModuleManager::Init() {
+
+        for (const auto &i: this->modules) {
+            spdlog::debug("\"{0}\" is registered", i.descriptor.id);
+        }
+
+        //Sort
+        this->SortModules();
+
+        spdlog::debug("Sorted order:");
+        for (const auto &i: this->modules) {
+            spdlog::debug("\"{0}\" is registered", i.descriptor.id);
+        }
+
+        //Load
+        for (auto &i: this->modules) {
+            spdlog::info("Loading {0}", i.descriptor.id);
+            if (std::any_of(ITERATE(this->assemblies), [i](Assembly a) {
+                return a.id == i.descriptor.assembly && a.lib != nullptr;
+            })) {
+                spdlog::debug("✅ Assembly \"{0}\" is already loaded", i.descriptor.assembly);
+            } else {
+                spdlog::info("➕ Loading assembly \"{0}\"", i.descriptor.assembly);
+                this->LoadAssembly(i.descriptor.assembly);
+            }
+
+            this->LoadModule(i);
+            if (i.enabled)
+                spdlog::info("Module {0}({1}) created", i.module->GetName(), i.descriptor.id);
+        }
+
+
+
+        //PreInit
+        spdlog::info("Running PreInit");
+        for (auto &holder: this->modules) {
+            if (holder.enabled) {
+                holder.module->PreInit();
+            }
+        }
+
+        spdlog::info("Running PreInit");
+        for (auto &holder: this->modules) {
+            if (holder.enabled) {
+                holder.module->Init();
+            }
+        }
+
+        //Sort
+
+        this->finalized = true;
+    }
+
+    void ModuleManager::LoadAssembly(const std::string &id) {
+        auto assembly = std::find_if(ITERATE(this->assemblies), [&](const auto &item) {
+            return item.id == id;
+        });
+        if (assembly == this->assemblies.end()) {
+            spdlog::error("No such assembly: {0}", id);
+            return;
+        }
+
+
+        dylib *dllptr;
+
+        try {
+            dllptr = new dylib("./", assembly->path);
+        }
+        catch (std::exception &e) {
+            spdlog::error(e.what());
+            exit(1);
+        }
+
+        assembly->lib = dllptr;
+    }
+
+    void ModuleManager::dfs(const ModuleHolder &moduleHolder, std::vector<ModuleHolder> &sorted) {
```

Comment:
Glancing at this implementation, i feel like it's wrong. Can't quite put my finger on it.

It may be the fact that this is just an iteration, not a depth-first search.